Add private repository steps to Astro (#494)
astronomer/docs: software/customize-image.md

>
> For more detail on how to add Environment Variables both locally and on Astronomer, refer to our [Environment Variables doc](environment-variables.md).

## Install Python Packages from a Private GitHub Repository

This topic provides instructions for building your Astro project with Python packages from a private GitHub repository.  At a high level, this setup entails specifying your private packages in `requirements.txt`, creating a custom Docker image that mounts a GitHub SSH key for your private GitHub repositories, and building your project with this Docker image.

Although this setup is based on GitHub, the general steps can be completed with any hosted Git repository.

:::info

The following setup has been validated only with a single SSH key. Due to the nature of `ssh-agent`, you might need to modify this setup when using more than one SSH key per Docker image.

:::

### Prerequisites

> To build from a private repository, you need:

Reviewer: ```suggestion
To install Python packages from a private GitHub repository on Astronomer Software, you need:
```

- The [Astronomer CLI](cli-quickstart.md).
- A [Software project](create-project.md).
- Custom Python packages that are [installable via pip](https://packaging.python.org/en/latest/tutorials/packaging-projects/).
- A private GitHub repository for each of your custom Python packages.

> - A [GitHub SSH Key](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent) authorized to access your private GitHub repo(s).

Reviewer: ```suggestion
- A [GitHub SSH Key](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent) authorized to access your private GitHub repo(s).

This setup assumes that each custom Python package is hosted within its own private GitHub repository. Installing multiple packages from a single private GitHub repository is not supported.
```

### Step 1: Specify the Private Repository in Your Project

> To add a Python package from a private repository to your Software project, specify the repository's SSH URL in your project's `requirements.txt` file. This URL should be formatted as `git+ssh://git@github.com/<organization-name>/<repository>.git`.

Reviewer: ```suggestion
To add a Python package from a private repository to your Software project, specify the repository's SSH URL in your project's `requirements.txt` file. This URL should be formatted as `git+ssh://git@github.com/<your-github-organization-name>/<your-private-repository>.git`.
```

For example, to install the `mypackage1` & `mypackage2` from `myorganization`, as well as `numpy v 1.22.1`, you would add the following to `requirements.txt`:

```
git+ssh://git@github.com/myorganization/mypackage1.git
git+ssh://git@github.com/myorganization/mypackage2.git
numpy==1.22.1
```

> 

Reviewer: ```suggestion

This example assumes that the name of each of your Python packages is identical to the name of its corresponding GitHub repository. In other words,`mypackage1` is both the name of the package and the name of the repository.

```

### Step 2. Create Dockerfile.build

1. In your Software project, create a duplicate of your `Dockerfile` named `Dockerfile.build`.

2. In `Dockerfile.build`, add `AS stage` to the `FROM` line which specifies your Astronomer Certified image. For example, if you use Certified 2.2.5, your `FROM` line would be:

   ```text

>    quay.io/astronomer/ap-airflow:2.2.5-onbuild AS stage1

Reviewer 2: Consider adding `FROM` back to the start of this line

   ```

  :::caution

>   If you use an `-onbuild` distribution of Certified, you need to replace it with the more customizable base distribution before building your project from a private registry. This base distribution is built to be customizable and does not include default build logic. For more information, see [Distributions](ac-support-policy.md#distribution)

Reviewer: ```suggestion
  If you use the default distribution of Astronomer Certified, make sure to remove the `-onbuild` part of the image. The Astronomer Certified distribution that does not specify `-onbuild` is built to be customizable and does not include default build logic. For more information, see [Distributions](ac-support-policy.md#distribution)
```

  :::

3. In `Dockerfile.build` after the `FROM` line specifying your Certified image, add the following configuration:

    ```docker
    LABEL maintainer="Astronomer <humans@astronomer.io>"
    ARG BUILD_NUMBER=-1
    LABEL io.astronomer.docker=true
    LABEL io.astronomer.docker.build.number=$BUILD_NUMBER
    LABEL io.astronomer.docker.airflow.onbuild=true
    # Install Python and OS-Level Packages
    COPY packages.txt .
    RUN cat packages.txt | xargs apk add --no-cache

    FROM stage1 AS stage2
    RUN --mount=type=ssh,id=github apk add --no-cache --virtual .build-deps \
        build-base \
        git \
        python3 \
        openssh-client \
    && mkdir -p -m 0600 ~/.ssh && ssh-keyscan github.com >> ~/.ssh/known_hosts \
    # Install Python Packages
    COPY requirements.txt .
    RUN pip install --no-cache-dir -q -r requirements.txt

    FROM stage1 AS stage3
    # Copy requirements directory
    COPY --from=stage2 /usr/lib/python3.9/site-packages/ /usr/lib/python3.9/site-packages/
    COPY . .
    ```

    In order, these commands:

>     - Complete the standard installation of OS-level packages in `packages.txt`.
>     - Securely mount your SSH key during build, which ensures that the key itself is not stored in the resulting Docker image filesystem or metadata.
>     - Install Python-level packages from your private repository as specified in your `requirements.txt` file.

Reviewer: ```suggestion
    - Install any OS-level packages specified in `packages.txt`.
    - Securely mount your GitHub SSH key at build time. This ensures that the key itself is not stored in the resulting Docker image filesystem or metadata.
    - Install all Python-level packages in `requirements.txt` file, including those from a private GitHub repository.
```

Although this setup is based on GitHub, the general steps can be completed with any hosted Git repository.

:::info

The following setup has been validated only with a single SSH key. Due to the nature of `ssh-agent`, you might need to modify this setup when using more than one SSH key per Docker image.

:::

### Prerequisites

To install Python packages from a private GitHub repository on Astronomer Software, you need:

- The [Astronomer CLI](cli-quickstart.md).
- A [Software project](create-project.md).
- Custom Python packages that are [installable via pip](https://packaging.python.org/en/latest/tutorials/packaging-projects/).
- A private GitHub repository for each of your custom Python packages.
- A [GitHub SSH Key](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent) authorized to access your private GitHub repositories.

> This setup assumes that each custom Python package is hosted within its own private GitHub repository. Installing multiple packages from a single private GitHub repository is not supported.

Reviewer: @stuart23 Clarifying sentence 1

### Step 1: Specify the Private Repository in Your Project

To add a Python package from a private repository to your Software project, specify the repository's SSH URL in your project's `requirements.txt` file. This URL should be formatted as `git+ssh://git@github.com/<your-github-organization-name>/<your-private-repository>.git`.

For example, to install the `mypackage1` & `mypackage2` from `myorganization`, as well as `numpy v 1.22.1`, you would add the following to `requirements.txt`:

```
git+ssh://git@github.com/myorganization/mypackage1.git
git+ssh://git@github.com/myorganization/mypackage2.git
numpy==1.22.1
```

> This example assumes that the name of each of your Python packages is identical to the name of its corresponding GitHub repository. In other words,`mypackage1` is both the name of the package and the name of the repository.

Reviewer: @stuart23 Clarifying sentence 2
